src/Java.Interop.BootstrapTasks/Java.Interop.BootstrapTasks/JdkInfo.cs

```diff
@@ -14,74 +17,265 @@ public class JdkInfo : Task
 
 		public override bool Execute ()
 		{
 			Log.LogMessage (MessageImportance.Low, $"Task {nameof (JdkInfo)}");
-			Log.LogMessage (MessageImportance.Low, $"  {nameof (Output)}: {Output}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (JdksRoot)}: {JdksRoot}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (MakeFragmentFile)}: {MakeFragmentFile}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (MaximumJdkVersion)}: {MaximumJdkVersion}");
+			Log.LogMessage (MessageImportance.Low, $"  {nameof (PropertyFile)}: {PropertyFile}");
+
+			var maxVersion  = GetMaxJdkVersion ();
+			var java_home   = GetJavaHomePathFromEnvironment ();
+			if (java_home != null) {
+				var java_home_v = GetVersionFromPath (java_home);
+				if (maxVersion != null && java_home_v != null && java_home_v > maxVersion) {
+					Log.LogMessage (MessageImportance.Low, $"  Skipping JAVA_HOME default value of `{java_home}` as it exceeds MaximumJdkVersion={MaximumJdkVersion}.");
+					java_home = null;
+				}
+				if (java_home != null && !Directory.Exists (Path.Combine (java_home, "include"))) {
+					Log.LogMessage (MessageImportance.Low, $"  Skipping JAVA_HOME default value of `{java_home}` as it does not contain an `include` subdirectory.");
+					java_home = null;
+				}
+			}
+			java_home = java_home ?? GetJavaHomePathFromMachine (maxVersion);
 
-			var javaSdkPath = GetJavaSdkPath ();
-			if (string.IsNullOrEmpty(javaSdkPath)) {
-				Log.LogError ("JavaSdkPath is blank");
+			if (string.IsNullOrEmpty (java_home)) {
+				Log.LogError ("Could not determine JAVA_HOME location. Please set JdksRoot or export the JAVA_HOME environment variable.");
 				return false;
 			}
 
-			Log.LogMessage (MessageImportance.Low, $"  JavaSdkPath: {javaSdkPath}");
+			var jarPath     = "";
+			var javacPath   = "";
+			var jdkJvmPath  = "";
+			var includes    = new List<string> ();
+
+			if (Path.DirectorySeparatorChar == '\\') {
+				jarPath     = Path.Combine (java_home, "bin", "jar.exe");
+				javacPath   = Path.Combine (java_home, "bin", "javac.exe");
+				jdkJvmPath  = Path.Combine (java_home, "jre", "bin", "server", "jvm.dll");
+			} else {
+				jarPath     = Path.Combine (java_home, "bin", "jar");
+				javacPath   = Path.Combine (java_home, "bin", "javac");
+				if (OnMacOS) {
+					jdkJvmPath  = Directory.EnumerateFiles (java_home, "libjli.dylib", SearchOption.AllDirectories)
+						.FirstOrDefault ();
+				} else {
+					jdkJvmPath  = Path.Combine (java_home, "jre", "bin", "server", "libjvm.so");
+				}
+			}
 
-			var jvmPath = Path.Combine (javaSdkPath, "jre", "bin", "server", "jvm.dll");
-			if (!File.Exists (jvmPath)) {
-				Log.LogError ($"JdkJvmPath not found at {jvmPath}");
+			if (!File.Exists (jarPath)) {
+				Log.LogError ("Could not determine location `jar`.");
+				return false;
+			}
+			if (!File.Exists (javacPath)) {
+				Log.LogError ("Could not determine location `javac`.");
+				return false;
+			}
```

Comment:
For the sake of staying DRY I'd put the `File.Exists` in a method (local or not) returning bool and just use

    if (!FileExists (jarPath))
         return false;

`FileExists` can construct error message from the path to get the missing binary name